src/peft/utils/save_and_load.py

```diff
@@ -231,23 +233,35 @@ def renamed_dora_weights(k):
             )
 
     # DEAL WITH EMBEDDINGS
-    # check the common embedding layers in `target_modules` to reset `save_embedding_layers` if necessary
-    is_embedding_in_target_modules = False
+    #
+    # save_embedding_layer="auto" needs to check the following logic:
+    #
+    # - when vocab size was NOT changed, embeddings should be saved only when targeted
+    # but not when
+    # - using PeftType.TRAINABLE_TOKENS
+    # - LoRA using trainable_token_indices (since their goal is to space-efficient)
+    # but
+    # - when vocab size was changed, embeddings should be saved automatically regardless to cover this
+    #   scenario: 1) fine-tune embedding, 2) resize embedding, 3) train with trainable tokens
+    #
     embedding_is_targeted = False
     if hasattr(config, "target_modules"):
-        if isinstance(config.target_modules, str):
-            # TODO: implement this; note: this change is not directly related to the PR, the bug already existed b4
-            pass
+        if isinstance(config.target_modules, str) and config.target_modules != INCLUDE_LINEAR_LAYERS_SHORTHAND:
+            embedding_is_targeted = any(
+                match_target_against_key(config.target_modules, k)
+                for k, _ in model.get_base_model().named_modules()
+                if any(e in k for e in EMBEDDING_LAYER_NAMES)
```

Comment:
Just wondering if `e in k` might be a bit too vague as a match, perhaps `e + "." in k`?